### DISCUSSION TITLE
[Extension dev] Custom option for .ideavimrc

### DISCUSSION THREAD

**Discussion by MunsterPlop** [JetBrains/ideavim#288] [General]
Hi,

I'm currently contributing to https://github.com/Mishkun/ideavim-sneak and I'm wondering if it's possible to create a custom option for `.ideavimrc` like `let g:sneak#option = 1` .
I've been looking around the codebase, had a look at `OptionsManager` and stuff, but I'm not sure how I should proceed.

Thanks in advance for the help.

**Comment by AlexPl292** [comment]
Hi! This is not an option, but rather a variable. So, check out `VimScriptGlobalEnvironment` and it's usages. This thing is pretty simple and primitive, but it satisfies the requirements so far.

Example of usage: `com.maddyhome.idea.vim.extension.argtextobj.VimArgTextObjExtension#bracketPairsVariable`. You don't have to "register" this variable or something like that, just `get` it when you need it.

**Reply by MunsterPlop** [reply]
Thank you very much for your reply.

I looked at the different usages and tried using `VimScriptGlobalEnvironment.getInstance()` but `getVariables()` doesn't return anything even though I set `let g:sneak#myoption = 1` in `.ideavimrc`. I doubt I need to parse .ideavimrc myself, so
am I missing something here?

**Reply by AlexPl292** [reply]
You definitely don't have to parse `.ideavimrc` by yourself. Could you please try to remove `#` from your option name or remove whitespaces around `=`? I'll check what's happening right after I finish my current tasks.

**Reply by MunsterPlop** [reply]
I tried all of these with and without spaces around `=`:
```
let g:sneak#myoption=1
let g:sneakmyoption=1
let sneakmyoption=1
```
Still nothing returned.
Thank you for your time.

**Reply by AlexPl292** [reply]
Okay, I've checked it.
The thing with `#` fails, I'm not sure that we can support it now. However, `let g:sneakmyoption=1` works pretty good for me. How are you trying to get it from `VimScriptGlobalEnvironment`? You should use it's full name, I mean `...get("g:sneakmyoption")`.
If it doesn't work for you with full name `g:sneakmyoption`, could you please post a small gist with your code or something so I could check it?

**Reply by MunsterPlop** [reply]
I should mention I'm using Kotlin.

I tried it again with `let g:sneaksnext=1`:

```kotlin
   override fun init() {
        // Get the sneak options
        val env = VimScriptGlobalEnvironment.getInstance()
        // val value = env.variables.getOrDefault("g:sneaksnext", 0)
        val value = env.variables.get("g:sneaksnext")

        println("env = $env")
        println("env.variables = " + env.variables)
        println("g:sneaksnext = $value")

        // ...
    }
```

This is what I get: 
```
env = com.maddyhome.idea.vim.ex.vimscript.VimScriptGlobalEnvironment@3e9f1fd2
env.variables = {}
g:sneaksnext = null
```

**Reply by AlexPl292** [reply]
I think I know the issue. Do you have the line with `let g:sneaksnext=1` after `set sneak` line? If yes, the line with your option was not yet parsed at the moment when sneak is initialized. But this is incorrect behaviour and you should update IdeaVim dependency from 0.61 to 0.66. However, the problem is that 0.66 version of IdeaVim has 2020.2+ compatibility and this plugin is 2020.1+.
So, you have two options:
- bump up the versions of IdeaVim dependency and IJ
- Make a note that line with the option should stand before `set sneak`

**Reply by MunsterPlop** [reply]
That was indeed the problem.
I'm gonna create an issue about this and let the maintainer know.
Thank you very much for your help and patience.

Have a good day!
